vch/src/vch.c

```
@@ -15,34 +16,304 @@
 
 #include "vch.h"
 
-#define RD_PROD(h) ((h)->read.s->prod)
-#define RD_CONS(h) ((h)->read.s->prod)
+#define RD_PROD(h) ((h)->read->prod)
+#define RD_CONS(h) ((h)->read->cons)
+#define WR_PROD(h) ((h)->write->prod)
+#define WR_CONS(h) ((h)->write->cons)
 #define ORD10_RING_SZ (1 << 10)
 #define ORD11_RING_SZ (1 << 11)
-#define RD_RING_SZ(h) (())
+#define RD_RING_SZ(h) (1 << (h)->read_ord)
+#define WR_RING_SZ(h) (1 << (h)->write_ord)
+
+static void vch_wait(struct vch_handle *h)
+{
+	__ASSERT(h, "Invalid handler");
+	notify_evtchn(h->evtch);
+	k_sem_take(&h->sem, K_FOREVER);
+}
+
+static void vch_notify_cb(void *data)
+{
+	struct vch_handle *h = data;
+
+	__ASSERT(h, "Invalid handler");
+	k_sem_give(&h->sem);
+}
+
+static void vch_notify(struct vch_handle *h, int rw)
+{
+	uint8_t *target, val;
+
+	if (!h || !h->ring)
+		return;
+
+	dmb();
+	if (h->is_server)
+		target = &h->ring->srv_notify;
+	else
+		target = &h->ring->cli_notify;
+
+	val = __atomic_fetch_and(target, ~rw, __ATOMIC_SEQ_CST);
+	if (val & rw)
+		notify_evtchn(h->evtch);
+}
 
 int vch_open(domid_t domain, const char *path, size_t min_rs, size_t min_ws,
 	     struct vch_handle *h)
 {
-	return -EINVAL;
-}
+	int rc;
+	grant_ref_t ring_gref;
+	uintptr_t ring_pfn;
+	char xs_key_scratch[64] = { 0 }, xs_val_scratch[8] = { 0 };
+
+	__ASSERT(h, "Invalid handler");
```

AFAIR, you need to return an error in this case. `vch_open` is externally visible, so using `__ASSERT` in an input parameter is not a best idea. ASSERTs usually used for checking internal state of a module.

---

fixed in [8eb23b6](https://github.com/xen-troops/zephyr-xenlib/pull/20/commits/8eb23b6fb030f2017287c1f94e9f89507c316797)